feat: add ascii repr in the console

apkrelling opened the pull request:
## Pull Request description

Add basic ascii AST visualization function and quick example.

This aims to help solve #22 .


## How to test these changes:

run the `docs/tutorials/ascii_ast_example.ipynb` notebook

## Pull Request checklists

This PR is a:

- [ ] bug-fix
- [X] new feature
- [ ] maintenance

About this PR:

- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [X] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:

- [X] I have reviewed the changes and it contains no misspelling.
- [X] The code is well commented, especially in the parts that contain more
      complexity.
- [] New and old tests passed locally.

## Additional information
#### png visualization:
![Screenshot from 2024-03-06 11-40-39](https://github.com/arxlang/astx/assets/74330736/3fb2fdc2-1ab5-49e5-9c83-9349f78c1b09)
#### ascii visualization:
![ascii_ast](https://github.com/arxlang/astx/assets/74330736/b5d84f5b-a159-47ef-b8e8-5960c2336040)

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

xmnlab commented on src/astx/asciiviz.py:
```diff
@@ -0,0 +1,72 @@
+displayed directly in the terminal window.
+
+"""
+from typing import Dict, List, Tuple
```
```suggestion
from __future__ import annotations
```

`Dict, List, Tuple` are deprecated. Now, please import annotations from `__futures__` and annotate with these words in lower case, for example:

```python
def get_nodes_edges(
    dic: dict[str, str],
    nodes: list[str] = [],
    edges: list[str] = [],
) -> tuple[list[str], list[str]]:
   ...
   ```

apkrelling commented:
Hey @xmnlab , I've made the changes we discussed:
* for the type hints, used `from __future__ import annotations` and `from typing import Any`
* reduced the number of indentation levels as you suggested
* emptied the poetry.lock file so there wouldn't be conflits (or it would be easier to solve any conflicts)
* used nodes=list(nodes)

apkrelling commented:
Reopening the pull request.

apkrelling commented:
## Changes:
- removed the notebook example (to be updated and added later)
- looked into other forms of doing ascii representation - such as [dot-to-ascii](https://dot-to-ascii.ggerganov.com/) - but asciinet/graph_to_ascii was more suitable due to the tree layout
- `traverse_ast` func now is suitable for both list and dict ASTs
- overloaded `graph_to_ascii` func to be able to deal with repeated node names.


## How to test these changes:

```python
from astx.viz import traverse_ast, visualize, graph_to_ascii
import astx

block = astx.Block()
decl_a = astx.VariableDeclaration("a", type_=astx.Int32, value=astx.LiteralInt32(1))
decl_b = astx.VariableDeclaration("b", type_=astx.Int32, value=astx.LiteralInt32(1))
var_a = astx.Variable("a")
var_b = astx.Variable("b")
sum_op = astx.BinaryOp(op_code="+", lhs=var_a, rhs=var_b)
block.append(decl_a)
block.append(decl_b)
block.append(sum_op)
visualize(block.get_struct(simplified=True)) # png viz

graph = traverse_ast(block.get_struct(simplified=True))
graph = graph_to_ascii(graph) # ascii viz
```

## Pull Request checklists

This PR is a:

- [ ] bug-fix
- [X] new feature
- [ ] maintenance

About this PR:

- [X] it includes tests.
- [X] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [X] pre-commit hooks were executed locally.
- [X] this PR requires a project documentation update.

Author's checklist:

- [X] I have reviewed the changes and it contains no misspelling.
- [X] The code is well commented, especially in the parts that contain more
      complexity.
- [X] New and old tests passed locally.

## Additional information

#### ascii visualization:
![Screenshot from 2024-05-13 13-47-18](https://github.com/arxlang/astx/assets/74330736/c3392a8b-480e-496e-89ca-24205cd6c1df)

## Reviewer's checklist

Copy and paste this template for your review's note:

```
## Reviewer's Checklist

- [ ] I managed to reproduce the problem locally from the `main` branch
- [ ] I managed to test the new changes locally
- [ ] I confirm that the issues mentioned were fixed/resolved .
```

xmnlab reviewed (changes requested):
hey @apkrelling, thanks for working on that!

I think that it is almost ready to go :)

some comments:

- the asciinet should be added by poetry: `poetry add "asciinet@>=0.3.1"` ... you maybe don't need it inside conda/dev.yaml .. just add it there if the CI is raising the issue with distlib ... in that case you need it both added by poetry and inside the conda/dev.yaml
- as you are importing msgpack explicitely inside viz.py, so you need to add it with poetry as well: `poetry add "msgpack@>=1"`
- call your new function here: https://github.com/arxlang/astx/blob/main/src/astx/base.py#L161

let me know if you have any questions or have any other comment, please

apkrelling commented:
## Update:
- added asciinet and msgpack to poetry
- turned `traverse_ast` into `traverse_ast_ascii` and `traverse_ast_png`
- added the `graph_to_ascii` func in [base.py](https://github.com/arxlang/astx/blob/main/src/astx/base.py#L161)
- changed the return of `graph_to_ascii_overload` from `None` to `str`.


## How to test these changes:

```python
from astx.viz import traverse_ast, visualize, graph_to_ascii
import astx

block = astx.Block()
decl_a = astx.VariableDeclaration("a", type_=astx.Int32, value=astx.LiteralInt32(1))
decl_b = astx.VariableDeclaration("b", type_=astx.Int32, value=astx.LiteralInt32(1))
var_a = astx.Variable("a")
var_b = astx.Variable("b")
sum_op = astx.BinaryOp(op_code="+", lhs=var_a, rhs=var_b)
block.append(decl_a)
block.append(decl_b)
block.append(sum_op)
visualize(block.get_struct(simplified=True)) # png viz

dot_graph = traverse_ast(block.get_struct(simplified=True))
graph = graph_to_ascii(dot_graph) # ascii viz
print(graph)
```

xmnlab reviewed (approved):
thanks for working on that @apkrelling ! appreciate that
